### PR TITLE
Postgres master slave

### PR DESCRIPTION
**Pull request by pallinger**
master/slave replication support, tested on Debian 10 (should work on RedHat too, but not tested), and postgres 11 (should work for up to postgres 13; postgres 14 changed replication config syntax).

- `db.postgres.replication.enabled` variable has to be true (false by default)
- `db.postgres.replication.servers` must have a list of ip addresses to trigger the appropriate changes to configure pg_hba.conf, postgresql.conf and firewall.
- `firewall=ufw` must also be defined to configure an ufw firewall
- The other changes are to allow the config files to be edited/placed properly on Debian as well as on RedHat.
- There is no actual replication initialized (i.e. copying the database and setting up recovery.conf), that must be done manually.

### CODE REVIEW

File: templates/pg_hba.conf.j2 [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# "local" is for Unix domain socket connections only
+local   all             all                                     trust
```
Comment (donsizemore) [inline templates/pg_hba.conf.j2:2]: is there a reason this needs to be trust, rather than ident?

File: files/pg_hba.conf [hunk 2]
```diff
@@ -1,6 +0,0 @@
-# "local" is for Unix domain socket connections only
```
Comment (donsizemore) [inline files/pg_hba.conf:1]: is files/pg_hba.conf extraneous, since we'd be placing a template?

Comment (pallinger) [reply files/pg_hba.conf:1]: Yes, I think it is extraneous.

**Comment by donsizemore** [comment]
@pallinger Overall this looks great and I appreciate the functionality. My only concern is that the templated pg_hba.conf leaves PG auth as "trust" and "password" rather than "peer" and "scram-sha-256" as Postgres now defaults in RedHat-land. Are these settings necessary for replication?

**Comment by pallinger** [comment]
I do not think so, I will try to tweak the parameters. We do not actually need them, as there is a wireguard tunnel between the master and the slave, but better encrypt twice than none.

File: files/pg_hba.conf (hunk 2 above)
Comment (pallinger) [reply files/pg_hba.conf:1]: It seems that files/pg_hba.conf is already removed in my branch.

**Comment by pallinger** [comment]
The "local   all             all                                     trust " and "host    all             all             127.0.0.1/32            password" row is verbatim from the original pg_hba.conf in the current gdcc/develop branch.

The first allows login by the local "postgres" user locally using unix socket with no password. This is useful, and no security risk as that user already has write access to all postgres files.

The second allows local users, for example dataverse itself, to connect using username-password. This is required for dataverse currently, AFAIK. It poses no security risk as to sniff the loopback network interface, you need root access anyway.

As for the actual replication entries, I can change md5 to scram-sha-256, although not strictly necessary:

https://www.postgresql.org/docs/current/auth-password.html
"To ease transition from the md5 method to the newer SCRAM method, if md5 is specified as a method in pg_hba.conf but the user's password on the server is encrypted for SCRAM (see below), then SCRAM-based authentication will automatically be chosen instead."

I tried that, and realized that then you must change the password_encryption setting in postgresql.conf, and re-set all passwords in the system.

I think leaving it md5 is better for existing installed systems.

I will shortly push some clean-up for the code.

**Comment by pallinger** [comment]
So the code is somewhat cleaned up, and ready to go, I hope.

**Comment by pallinger** [comment]
Can anyone review this?

**Comment by donsizemore** [comment]
> Can anyone review this?

I'm afraid I'm swamped with my day job responsibilities, and haven't had the spare cycles to devote to this repository as I'd like. I'm the sole admin of the repo; would you like to assume a co-administrator role on the project?

**Comment by pallinger** [comment]
> > Can anyone review this?
> 
> I'm afraid I'm swamped with my day job responsibilities, and haven't had the spare cycles to devote to this repository as I'd like. I'm the sole admin of the repo; would you like to assume a co-administrator role on the project?

I have looked at the pull requests of this repo this year, and it seems it had only 5 PRs that were not me or you, so it am happy to assume a co-admin role.

**Comment by pallinger** [comment]
Is there a set way to do QA for PRs? Some tests? CI? Just trying to run the code with some default settings?

**Comment by donsizemore** [comment]
I usually specify the repo and branch in a copy of https://github.com/GlobalDataverseCommunityConsortium/dataverse-ansible/blob/develop/tests/group_vars/jenkins.yml which runs IQSS' unit and integration tests. Vagrant VM, EC2 instance, however you like.

**Comment by pallinger** [comment]
Could this be merged? Or should I merge it?
It should be completely harmless if replication is not enabled...

**Comment by donsizemore** [comment]
@pallinger looks fine to me, though the UFW step seems outside of the scope of this role? (and will bomb on systems which don't use UFW?)

**Comment by pallinger** [comment]
> @pallinger looks fine to me, though the UFW step seems outside of the scope of this role? (and will bomb on systems which don't use UFW?)

It is guarded by a "firewall" variable that must be set to "ufw" value, so it would be of no consequence otherwise. But, if there is an ufw firewall, then it punches the needed hole for replication.

**Pull request merged**